test/metabase/test_runner/assert_exprs/approximately_equal.clj

```diff
@@ -0,0 +1,270 @@
+  compare things. This works inside collections as well.
+
+    (is (=? {:m #exactly {:a 1}}
+            {:m {:a 1, :b 2}}))
+    ;; =>
+    expected: {:m #exactly {:a 1}}
+
+      actual: {:m {:a 1, :b 2}}
+        diff: - {:m (not (= #exactly {:a 1} {:a 1, :b 2}))}
+              + nil
+
+  `#schema` compares things to a [[schema.core]] Schema:
+
+    (is (=? {:a 1, :b #schema {s/Keyword s/Int}}
+            {:a 1, :b {:c 2}}))
+    => ok
+
+    (is (=? {:a 1, :b #schema {s/Keyword s/Int}}
+            {:a 1, :b {:c 2.0}}))
+    =>
+    expected: {:a 1, :b #schema {(pred keyword?) (pred integer?)}}
+
+      actual: {:a 1, :b {:c 2.0}}
+        diff: - {:b {:c (not (integer? 2.0))}}
+              + nil
+
+  `#approx` compares whether two numbers are approximately equal:
+
+    ;; is the difference between actual and 1.5 less than ±0.1?
+    (is (=? #approx [1.5 0.1]
+            1.51))
+    => true
+
+    (is (=? #approx [1.5 0.1]
+            1.6))
+    =>
+    expected: #approx [1.5 0.1]
+
+      actual: 1.6
+        diff: - (not (approx= 1.5 1.6 #_epsilon 0.1))
+              + nil"
+  (:require
+   [clojure.algo.generic.math-functions :as algo.generic.math]
+   [clojure.pprint :as pprint]
+   [methodical.core :as methodical]
+   [schema.core :as s]))
+
+(set! *warn-on-reflection* true)
+
+(methodical/defmulti ^:dynamic =?-diff
+  "Multimethod to use to diff two things with `=?`. Despite not having earmuffs, this is dynamic so it can be rebound at
+  runtime."
+  {:arglists '([expected actual])}
+  (fn [expected actual]
+    [(type expected) (type actual)]))
+
+(defn- add-primary-methods
+  "Add primary methods in map `m` of dispatch value -> method fn to [[*impl*]]. Return a new multifn with those methods
+  added."
+  [m]
+  (reduce
+   (fn [multifn [dispatch-value f]]
+     (methodical/add-primary-method multifn dispatch-value f))
+   =?-diff
+   m))
+
+(def ^:dynamic *debug*
+  "Whether to enable Methodical method tracing for debug purposes."
+  false)
+
+(defn =?-diff*
+  "Are `expected` and `actual` 'approximately' equal to one another?"
+  ([expected actual]
+   (=?-diff* =?-diff expected actual))
+
+  ([diff-fn expected actual]
+   (let [diff-fn (if (map? diff-fn)
+                   (add-primary-methods diff-fn)
+                   diff-fn)]
+     (binding [=?-diff diff-fn]
+       (if *debug*
+         (methodical/trace diff-fn expected actual)
+         (diff-fn expected actual))))))
+
+;;;; Default method impls
+
+(methodical/defmethod =?-diff :default
+  [expected actual]
+  (when-not (= expected actual)
+    (list 'not= expected actual)))
+
+(methodical/defmethod =?-diff [Class Object]
+  [expected-class actual]
+  (when-not (instance? expected-class actual)
+    (list 'not (list 'instance? expected-class actual))))
+
+(methodical/defmethod =?-diff [java.util.regex.Pattern String]
+  [expected-regex s]
+  (when-not (re-find expected-regex s)
```

Comment:
I expected `re-matches` instead of `re-find`. I guess people have different expectations, but for me a pattern "matching" a string is more closely associated with being approximately equal than a pattern "occurring in" the string.

Reply:
good point. I will change it